### PR TITLE
[4.4] Use array values of app paths returned by App::path() for I18n commands

### PR DESCRIPTION
**Pull request by fabian-mcfly**
`App::path()` returns the config values for the `App.paths.X` config path. The returned array might be set as an associative array since there currently is no logic to prevent that.

I have a multi-client setup (with dynamic locale folders) and I like to keep track of existing files and more. So calling `App::path('locales')['customer']` returns an additional path I (dynamically) set in my config. Surely a one-in-a-million case but those changes do not break anything but help in case those paths have non-numeric keys.

### CODE REVIEW

**Comment by fabian-mcfly** [comment]
All other usages of `App::path()` already work with associative arrays. For example

https://github.com/cakephp/cakephp/blob/6aeb856ee523d326567d745a69db9ee7c9c7d72c/src/Core/ConventionsTrait.php#L143

is using the `current()` function instead of accessing `[0]`.

**Comment by ADmad** [comment]
Test cases should be added to show this change is required else it will be forgotten and someone will remove the array_values() calls in the future as redundant.

That said it's best to have your app specific data separate and not mix it with values used by the core.

**Comment by fabian-mcfly** [comment]
I will try my best to provide matching tests.


> That said it's best to have your app specific data separate and not mix it with values used by the core.

I understand. But since most of the core already works flawlessly with associative arrays, I thought it would be best to work with the values of those arrays where only the values matter. At least as long as there's no logic that ensures the values are basic arrays.

Another way would be to change `App::path()` so it will always return a basic array (calling `array_values` in there), but that might lead to unexpected behavior.

File: tests/TestCase/Command/I18nExtractCommandTest.php [hunk 1]
```diff
@@ -58,6 +64,8 @@ public function tearDown(): void
         $fs = new Filesystem();
         $fs->deleteDir($this->path);
         $this->clearPlugins();
+
+        Configure::write('App.paths', $this->configPaths);
```
Comment (markstory) [inline tests/TestCase/Command/I18nExtractCommandTest.php]: You shouldn't need to do this backup/restore. The `TestCase::setup()` and `teardown` methods already do this.

Comment (fabian-mcfly) [reply tests/TestCase/Command/I18nExtractCommandTest.php]: Makes sense, yes. I took it from here: https://github.com/cakephp/cakephp/blob/6aeb856ee523d326567d745a69db9ee7c9c7d72c/tests/TestCase/I18n/MessagesFileLoaderTest.php#L36-L49

I found a few other similar places. Shall I remove it there as well? Would that require a different pull request?

Comment (markstory) [reply tests/TestCase/Command/I18nExtractCommandTest.php]: > Shall I remove it there as well? Would that require a different pull request?

You can update the other backup/restore pattern usage if you want. It doesn't need to be a separate pull request.

Comment (fabian-mcfly) [reply tests/TestCase/Command/I18nExtractCommandTest.php]: Done!

File: tests/TestCase/Command/I18nCommandTest.php [hunk 2]
```diff
@@ -91,6 +92,37 @@ public function testInit(): void
...
+            mkdir($deDir, 0770, true);
+        }
+        file_put_contents($this->localeDir . 'default.pot', 'Testing POT file.');
+        file_put_contents($this->localeDir . 'cake.pot', 'Testing POT file.');
+        if (file_exists($deDir . 'default.po')) {
+            unlink($deDir . 'default.po');
+        }
+        if (file_exists($deDir . 'default.po')) {
+            unlink($deDir . 'cake.po');
+        }
```
Comment (ravage84) [inline tests/TestCase/Command/I18nCommandTest.php]:
```suggestion
        if (file_exists($deDir . 'cake.po')) {
            unlink($deDir . 'cake.po');
        }
```

File: tests/TestCase/Command/I18nExtractCommandTest.php [hunk 3]
```diff
@@ -392,4 +392,35 @@ public function testExtractWithInvalidPaths(): void
...
+            'templates' => ['customKey' => TEST_APP . 'templates' . DS],
+            'locales' => ['customKey' => TEST_APP . 'resources' . DS . 'locales' . DS],
+        ]);
+
+        $this->exec(
+            'i18n extract ' .
+            '--merge=no ' .
+            '--extract-core=no ',
+            [
+                '', //Sending two empty inputs so \Cake\Command\I18nExtractCommand::_getPaths() loops through all paths
```
Comment (ravage84) [inline tests/TestCase/Command/I18nExtractCommandTest.php]:
```suggestion
                // Sending two empty inputs so \Cake\Command\I18nExtractCommand::_getPaths()
                // loops through all paths
                '',
```

File: tests/TestCase/Command/I18nExtractCommandTest.php [hunk 4]
```diff
@@ -392,4 +392,35 @@ public function testExtractWithInvalidPaths(): void
         $expected = '#: ./tests/test_app/templates/Pages/extract.php:';
         $this->assertStringContainsString($expected, $result);
     }
+
+    /**
+     * Test with associative arrays in App.path.locales and App.path.templates.
+     * A simple
```
Comment (ravage84) [inline tests/TestCase/Command/I18nExtractCommandTest.php]: Sentence not finished?

**Review by ravage84 — CHANGES_REQUESTED** [review]

**Comment by ajibarra** [comment]
@fabian-mcfly could you please confirm you have implemented all changes requested by @ravage84 so it can be merged? thank you!

**Comment by fabian-mcfly** [comment]
@ajibarra Yep, I did.

**Pull request merged**